### PR TITLE
Ch01: add lemma for proof 4

### PR DESCRIPTION
**Pull request by rwst**
There might be improvements to prod_Icc_succ_div that I don't see. I also think this should go in mathlib to the other prod lemmata. What do you say, will you take it as is? If not, I'll try again.

### CODE REVIEW

File: FormalBook/Ch01_Six_proofs_of_the_infinity_of_primes.lean [hunk 1]
```diff
@@ -224,6 +224,33 @@ lemma H_P4_1 {k p: ℝ} (hk: k > 0) (hp: p ≥ k + 1): p / (p - 1) ≤ (k + 1) /
...
+    rw [← mul_prod_Ico_eq_prod_Icc, Nat.cast_succ]
+    rw [add_assoc, one_add_one_eq_two, prod_Ico_succ_top]
+    nth_rewrite 1 [mul_comm]
+    rw [prod_Ico_mul_eq_prod_Icc, hall, mul_div ((n : ℝ) + 1) ((n : ℝ) + 2) ((n : ℝ) + 1)]
+    rw [mul_comm, mul_div_cancel_of_imp]
+    assumption'
+    . intro hl
+      have : ¬((n : ℝ) + 1 = 0) := by linarith
+      contradiction
+    . exact Nat.one_le_of_lt hn
```
Comment (mo271) [inline FormalBook/Ch01_Six_proofs_of_the_infinity_of_primes.lean:247]:
```suggestion
  rw [← Nat.Ico_succ_right]
  induction' n with n h
  · simp
  · rw [Finset.prod_Ico_succ_top <| Nat.le_add_left 1 n]
    norm_num
    cases' lt_or_ge n 2 with _ h2
    · interval_cases n
      · tauto
      · norm_num
    specialize h h2
    field_simp [Finset.prod_eq_zero_iff] at h ⊢
    rw [h]
    ring
```

Golfed it a little using the original ideas

File: FormalBook/Ch01_Six_proofs_of_the_infinity_of_primes.lean [hunk 2]
```diff
@@ -224,6 +224,33 @@ lemma H_P4_1 {k p: ℝ} (hk: k > 0) (hp: p ≥ k + 1): p / (p - 1) ≤ (k + 1) /
...
+
+lemma H_P4_2 (hx : x ≥ 3) (hpi3 : (π 3) = 2): (∏ x in Icc 1 (π x), ((x + 1) : ℝ) / x) = (π x) + 1 := by
```
Comment (mo271) [inline FormalBook/Ch01_Six_proofs_of_the_infinity_of_primes.lean:249]:
```suggestion
lemma H_P4_2 (hx : x ≥ 3) (hpi3 : (π 3) = 2) : (∏ x in Icc 1 (π x), ((x + 1) : ℝ) / x) = (π x) + 1 := by
```

**Review by mo271 — APPROVED** [review]
Thanks, this looks great!

**Comment by mo271** [comment]
> There might be improvements to prod_Icc_succ_div that I don't see. I also think this should go in mathlib to the other prod lemmata. What do you say, will you take it as is? If not, I'll try again.

Seems like there is `Finset.prod_Ico_succ_div_top` but genereally it seems like there is more lemmata for `Ico` than for `Icc` in mathlib. I don't know if this is by design or they are just missing and waiting to be added. Best ask at Zulip if you have a concrete suggestion.

File: FormalBook/Ch01_Six_proofs_of_the_infinity_of_primes.lean (hunk 1 above)
Comment (rwst) [reply FormalBook/Ch01_Six_proofs_of_the_infinity_of_primes.lean:247]: Great, thanks.

**Comment by mo271** [comment]
Thanks!

**Pull request merged**